--- focus_cami.py	2015-12-10 16:16:57.727059629 +0100
+++ focus_cami.py.new	2015-12-10 16:18:03.228977531 +0100
@@ -72,8 +72,8 @@
     print "Jellyfish not installed. Please download it at http://www.cbcb.umd.edu/software/jellyfish"    
     
 #check if the query exists or if it is adding genome to the db
-elif os.path.isfile(parameters["-q"])!=True:
-    print "Please use -q and select a valid query!"
+#~ elif os.path.isfile(parameters["-q"])!=True:
+    #~ print "Please use -q and select a valid query!"
 
 #check if jellyfish is installed
 elif which("jellyfish")==None:
@@ -209,7 +209,7 @@
                     c+=1
 
                 #Writes tabular output!
-                o=open(parameters["-q"]+"__output.txt","w+")
+                o=open("__output.txt","w+")
                 o.write("Query: "+parameters["-q"]+"\n")
                 o.write("K-mer size: "+parameters["-k"]+"\n\n")
                 for result in tabular:
