--- ../../dockerruns/phylosift_1/NewickParser.pm	2015-11-27 21:02:07.141044669 +0100
+++ NewickParser.pm	2015-11-27 21:03:56.790643104 +0100
@@ -154,7 +154,15 @@
             $self->_end('nhx_tag');
 
             $token = next_token(\$newick, ",);"); #move to , or )
-        }
+        } elsif ($token =~ /\[/) {
+            # This is a hack to make AMPHORA2 work
+            if ($token =~ /\[(\S+)\]/) {
+                $self->_start('bootstrap');
+                $self->_chars($1);
+                $self->_end('bootstrap');
+	    }
+            $token = next_token(\$newick, ",);");        
+	}
         $state = 5;
       } elsif ($state == 5) { # end node
         if($token eq ')') {
